src/http/common.cc

```diff
@@ -124,6 +124,46 @@ output_stream<char> make_http_chunked_output_stream(output_stream<char>& out) {
     return output_stream<char>(http_chunked_data_sink(out), 32000, opts);
 }
 
+class http_content_length_data_sink_impl : public data_sink_impl {
+    output_stream<char>& _out;
+    size_t _remaining_bytes;
+
+public:
+    http_content_length_data_sink_impl(output_stream<char>& out, size_t len) : _out(out), _remaining_bytes(len) {
+    }
+    virtual future<> put(net::packet data)  override { abort(); }
+    using data_sink_impl::put;
+    virtual future<> put(temporary_buffer<char> buf) override {
+        if (buf.size() == 0 || _remaining_bytes == 0) {
+            return make_ready_future<>();
+        }
+        auto size = std::min(buf.size(), _remaining_bytes);
+        return _out.write(buf.get(), size).then([this, size] {
+            _remaining_bytes -= size;
+        });
+    }
+    virtual future<> close() override {
+        if (_remaining_bytes != 0) {
+            return make_exception_future<>(std::runtime_error("partial request body write"));
```

Comment:
close() should never throw, it makes error handling impossible. @bhalevy has scars.

Perhaps we should fail the put() instead.

Reply:
Failing put() doesn't work. Consider this:

    out = make_http_content_length_output_stream(socket.output(), 20);
    co_await out.write("123");
    co_await out.flush();
    co_await out.close();

up until final close the sink has no idea there will be no more data coming.

Reply:
Ah, you're checking for underflow, not overflow.

Well, throwing on close is a no-no. Ask @bhalevy to show you his scars.